### ISSUE TITLE
[RFC] The choice between rich-text (html) vs markdown should be up to the author on an individual basis, not a monolithic attribute of the field

### ISSUE THREAD

**Issue by padawan** [bolt/bolt#4815]
TL;DR: The choice of format, for HTML-holding fields, should be left to the author on an individual entry basis.

If you want a practical editor for an HTML-holding field, right now Bolt forces you to chose between either html or Markdown as fieldtype. This means that, for _all entries_ of a specific type of content,  _everybody_ is forced into either a rich-text editor, or Markdown, or just a plain textarea. From experience, this choice should be left to the person editing the content — some prefer a rich-text editor, some prefer MD, some will choose depending on the content they're editing at one time.

Here's how it's done in Movable Type (you should study it, yes it's old and forgotten, but it's way more mature and flexible than many CMS out there):

![mt-wysiwyg](https://cloud.githubusercontent.com/assets/91673/12875709/07ecf83c-ce44-11e5-8f00-6cc1b2135c27.png)

Here, for any individual entry, I can pick between various formats: Rich-text (Bolt's "html"), Markdown (with or without SmartyPants), convert line breaks (paragraphs are separated with a blank line) or none (pure HTML, equivalent to Bolt's textarea). I can even change format mid-editing and see/edit the code.

![convert-line-breaks](https://cloud.githubusercontent.com/assets/91673/12875813/455a19b4-ce46-11e5-8475-e91f39c03e90.png)

![md](https://cloud.githubusercontent.com/assets/91673/12875815/4f697b84-ce46-11e5-9d0f-9b2ea0e05158.png)

Now if I switch from MD to Richtext, it converts MD to HTML:
![richtext](https://cloud.githubusercontent.com/assets/91673/12875817/55014d92-ce46-11e5-9b06-26bfbdbee520.png)

Here I can see and edit the HTML code directly (you'd get the same if you pick "none" for format):
![richtext-source-mode](https://cloud.githubusercontent.com/assets/91673/12875818/593599e0-ce46-11e5-8045-6a2778ae6b8f.png)

In essence, this would mean moving the field type into the individual content record rather than having just one type defined at the fieldtype level. And either allow for multiple formats (e.g. `type: html, markdown` for allowing rich-text and MD only), or creating a new flexible type that would allow a choice between all available formats when one edits such a field.

I can tell you from experience (with more than a hundred editors, journalists and web professionals) that they very much appreciate this kind of authorship freedom, especially with a heterogeneous bunch of technical and non-technical-savvy editors (IT folks and developer types overestimate the appeal of Markdown on the masses, it's just as ugly and incomprehensible to them as HTML).

**Comment by bobdenotter** [comment]
I have a feeling this should be an _extension_, and not be in core. Open to other opinions.

**Comment by Boorj** [comment]
Indeed, you can make your own extension that provides a new fieldtype with switch.

**Comment by SahAssar** [comment]
Yep, seems like extension material to me. @padawan hope you don't mind if I close this one? I think the forums might be a better place to get feedback on implementation of this as an extension. Feel free to reopen if there are unresolved issues here :)

**Comment by padawan** [comment]
Well, I think it should be in core, since it's in the core where you made that decision that a simple storage field that contains, in essence, HTML content (MD has no other purpose than to be converted to HTML) forces which editor is presented to the user. Basically, I'm saying "this is a body field, now let me, as an author, pick whichever format I'm confortable with". Your design choice is not about the field content but the type of editor it's displayed in. Do you see the difference and how it impacts both users and designers?

My use case is this: https://www.paris-web.fr/ — we have hundreds of pages, all with the same exact content model, but with a mix of HTML and Markdown depending on each page. Now I have to pick just one format, and force all our authors to edit all pages with just one format. The choice is not only Cornelian for a single developper who's going to make a decision impacting a dozen people, it feels backwards for those of us who are used to the possibility to switch freely between a rich-text editor, Markdown, or source.- depending on the content we're editing.

This said, I understand why you're so quickly dismissing this into plugin-land. I guess I'll have to dig deeper into the code and try to code something. Once you *see* it in action, you'll understand why it has to be in core ;-).

**Comment by SahAssar** [comment]
@padawan I'm sure that this will be very useful for a great number of people, but considering how easy it is to make a custom fieldtype in an extension and how much additional code/functionality this would probably require to be supported in core I'm not sure if I agree that it should be in core before it's "proven" and used.

Making a custom field from an extension is very easy, have a look at https://github.com/Raistlfiren/FAIconFieldType or one of the other fields in the extensions marketplace, it requires very little code to get something running.

Sorry if I get off as dismissive, not my intention at all, we just try to keep this tracker to things that fit within bolt core and that are actionable :)

**Comment by padawan** [comment]
Let's agree to disagree whether this ought to belong to the core or not. Now I have to learn how to code this, or convince someone else that it's useful :).

**Comment by GawainLynch** [comment]
While I could see the benefit in core, the bigger problem is resourcing.

We have a lot going on, and this section of the core team is not as big as the team working on functional code… and we all have real jobs too :grin: 

So adding all these things while nice, means someone has to develop it, someone has to write unit & acceptances tests for it (hard!), someone has to write documentation for it… and **then** someone has to be able to maintain maintain it.

**Comment by jadwigo** [comment]
And don't forget someone has to look at it from an UX perspective - one of the hardest things for end-users in drupal is the "why is my WYSIWYG text not displaying as I expect?"
